Integrate RMNRemote and OffRamp

RyanRHall opened the pull request:
## Motivation
Remove the requirement for self-transmitted RMN blessings

## Solution
Allow the commit DON to include RMN blessings at commitment time
This PR has a dependency on changes to [chainlink-ccip](https://github.com/smartcontractkit/chainlink-ccip/pull/84)

RyanRHall committed:
remove cursing from OffRamp

RyanRHall committed:
flatten MerkleRoot struct

RyanRHall committed:
add onrampAddress to MerkleRoot struct

RyanRHall committed:
write IRMNRemote, move MerkleRoot definition

RyanRHall committed:
swap IRMN for IRMNRemote in OffRamp

RyanRHall committed:
implement call to RMNemote#verify() from OffRamp

RayXpub commented on contracts/src/v0.8/ccip/interfaces/IRMNRemote.sol:
```diff
@@ -0,0 +1,22 @@
+  uint64 minSeqNr; // Minimum sequence number, inclusive
+  uint64 maxSeqNr; // Maximum sequence number, inclusive
+  bytes32 merkleRoot; // Merkle root covering the interval & source chain messages
+}
```
nit: missing struct packing comments

RyanRHall replied:
we had a discussion on struct packing this with research. It's not a storage struct so the benefits are minimal (though I think it would reduce tx size and probably some gas usage if we could struct pack it). Research doesn't want to change the order of fields because it would require changes to RMN, which I didn't feel strongly about pushing them to do.

I know you only mentioned the struct comments, but my thinking is that since this really isn't struct packed, maybe we should avoid the lines / comments that make it look like we tried to struct pack it lol.

RayXpub replied:
yeah no strong opinion here it was really a nit mainly for consistency across the codebase

RyanRHall replied:
uhg didn't realize this before but the lack of struct packing triggers solhint warnings (which is a good thing) but I hate the idea of adding solhint-ignore comments or increasing our warning count because we're too lazy to re-order a stuct :facepalm:. So I'm going to fix this after-all. cc @gtklocker

RensR replied:
Imo packing every single struct just makes sense. It's almost zero effort and just looks better. No need to reason about if it's ever going to be used in storage or anything :)

RayXpub commented on contracts/src/v0.8/ccip/interfaces/IRMNRemote.sol:
```diff
@@ -0,0 +1,22 @@
+/// @dev RMN depends on this struct, if changing, please notify the RMN maintainers.
+struct MerkleRoot {
+  uint64 sourceChainSelector; // Remote source chain selector that the Merkle Root is scoped to
+  bytes onrampAddress; // generic, to support arbitrary sources; for EVM2EVM, use abi.encodePacked
```
nit: casing + name is not consistant with other instances https://github.com/smartcontractkit/ccip/blob/9c2dc314f7a637776e23b8282ec248b2e06f8764/contracts/src/v0.8/ccip/offRamp/OffRamp.sol#L95

RyanRHall committed:
rename onrampAddress => onRampAddress

RyanRHall committed:
reorder fields in MerkleRoot struct

RyanRHall committed:
solhint tidy

RyanRHall committed:
solhint ignore struct packing issue

RyanRHall committed:
move RMNHome and RMNRemote into rmn/; integrate IRMNRemote

RyanRHall committed:
add RMNRemote to gethwrapper generation

RyanRHall committed:
remove RMN from OnRamp

RyanRHall committed:
add backwards-compatible cursing to RMNRemote contract

RyanRHall committed:
update solhint ignore

RyanRHall commented on contracts/src/v0.8/ccip/rmn/RMNRemote.sol:
```diff
@@ -146,6 +135,17 @@ contract RMNRemote is Ownable2Step, ITypeAndVersion {
   }
 
+  /// @notice If there is an active global or legacy curse, this function returns true.
+  function isCursed() external view returns (bool) {
+    return false; // XXX temporary workaround
+  }
+
+  /// @notice If there is an active global curse, or an active curse for `subject`, this function returns true.
+  /// @param subject To check whether a particular chain is cursed, set to bytes16(uint128(chainSelector)).
+  function isCursed(bytes16 subject) external view returns (bool) {
+    return false; // XXX temporary workaround
+  }
```
this will all be properly implemented in a follow-up PR

RyanRHall commented on contracts/.solhintignore:
```diff
@@ -43,5 +43,4 @@
-./src/v0.8/ccip/RMNRemote.sol
-./src/v0.8/ccip/RMNHome.sol
\ No newline at end of file
+./src/v0.8/ccip/rmn
```
Will remove in follow up PR

RyanRHall commented on contracts/package.json:
```diff
@@ -108,7 +108,7 @@
     "prettier": "^3.2.5",
     "prettier-plugin-solidity": "^1.3.1",
-    "solhint": "^5.0.1",
+    "solhint": "^5.0.3",
```
upgraded while attempting to debug [this](https://github.com/protofire/solhint/issues/597). Figured I'd leave it, but not a critical change.

RensR replied:
Generally we do npm bumps in the main repo and then merge back into this one. I've already bumped this in https://github.com/smartcontractkit/chainlink/pull/14093/files#diff-8fdb4aff7a0e74ab6d27fe514bc7e0baaa77fcf61286ae7ce7ef69f7aabfa23b

Maybe easier to reduce conflicts and undo this one

RyanRHall replied:
ah right. Will undo.

RyanRHall commented on contracts/src/v0.8/ccip/interfaces/IRMNRemote.sol:
```diff
@@ -0,0 +1,29 @@
+interface IRMNRemote {
+  function verify(MerkleRoot[] memory merkleRoots, Signature[] memory signatures) external view;
+
+  /// @notice If there is an active global or legacy curse, this function returns true.
+  function isCursed() external view returns (bool);
+
+  /// @notice If there is an active global curse, or an active curse for `subject`, this function returns true.
+  /// @param subject To check whether a particular chain is cursed, set to bytes16(uint128(chainSelector)).
+  function isCursed(bytes16 subject) external view returns (bool);
```
Question: are both of these required for backwards compatibility? I know `isCursed()` is used in the `Router`, so we definitely need that one. `isCursed(bytes16 subject)` I'm not sure about.

RyanRHall commented on contracts/src/v0.8/ccip/interfaces/IRMNRemote.sol:
```diff
@@ -0,0 +1,29 @@
+interface IRMNRemote {
+  function verify(MerkleRoot[] memory merkleRoots, Signature[] memory signatures) external view;
+
+  /// @notice If there is an active global or legacy curse, this function returns true.
+  function isCursed() external view returns (bool);
+
+  /// @notice If there is an active global curse, or an active curse for `subject`, this function returns true.
+  /// @param subject To check whether a particular chain is cursed, set to bytes16(uint128(chainSelector)).
+  function isCursed(bytes16 subject) external view returns (bool);
+}
```
Question: are both of these required for backwards compatibility? I know isCursed() is used in the Router, so it seems we need that one. isCursed(bytes16 subject) I'm not sure about.

RayXpub replied:
I think it `isCursed(bytes16 subject)` would be useful to be able to curse a specific lane.

RensR replied:
Both are required, the bytes16 version is used on 1.5 pools

RyanRHall commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -83,9 +82,12 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
   /// @notice Struct that contains the static configuration
   /// @dev RMN depends on this struct, if changing, please notify the RMN maintainers.
+  /// @dev not sure why solhint complainins about this, seems like a buggy detector
+  /// https://github.com/protofire/solhint/issues/597
+  // solhint-disable-next-line gas-struct-packing
```
this is fixable without the `solhint-disable` by reordering (see example in the linked issue), but I didn't want to change our code because of a linter bug.

RayXpub commented on contracts/src/v0.8/ccip/interfaces/IRMNRemote.sol:
```diff
@@ -0,0 +1,29 @@
+  uint64 minSeqNr; //              | Minimum sequence number, inclusive
+  uint64 maxSeqNr; // ─────────────╯ Maximum sequence number, inclusive
+  bytes32 merkleRoot; //             Merkle root covering the interval & source chain messages
+  bytes onRampAddress; //            generic, to support arbitrary sources; for EVM2EVM, use abi.encode
```
```suggestion
  bytes onRampAddress; //            Generic onramp address, to support arbitrary sources; for EVM2EVM, use abi.encode
```

RensR replied:
> EVM2EVM

EVM not EVM2EVM probably?

RensR replied:
not fixed yet (or fix not pushed?)

RayXpub commented on contracts/src/v0.8/ccip/interfaces/IRMNRemote.sol:
```diff
@@ -0,0 +1,29 @@
+struct Signature {
+  bytes32 r;
+  bytes32 s;
+}
```
nit: missing comments

RayXpub commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
I think we can also remove this one? https://github.com/smartcontractkit/ccip/blob/d1f5baf529b24e87c8a0dbc1955973b465768e15/contracts/src/v0.8/ccip/offRamp/OffRamp.sol#L136-L140

RayXpub commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -682,10 +655,6 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
     uint256 proofFlagBits
   ) internal view virtual returns (uint256 timestamp) {
```
q: do we want to keep this internal function? feels like it can be inlined now.

RensR replied:
+1 this is a single line

RyanRHall replied:
the `_verify()` function is overwritten in tests to support mocking verification results. Changing this would be a big reafactor of the test suite and would probably increase code complexity there. Opting to leave this as is unless you guys feel strongly.

RayXpub replied:
It feels a bit wrong to have a one liner internal function just because of tests. We can punt to another PR but I'd be in favor of removing it.

RyanRHall replied:
> It feels a bit wrong to have a one liner internal function just because of tests. We can punt to another PR but I'd be in favor of removing it.

I'll spend a little more time on it today and see how annoying it will be in tests. I was admittedly being a touch lazy yesterday 🙃.

RensR replied:
FYI easy solve: add a function in the offRampHelper that just sets `s_roots[sourceChainSelector][root]` to whatever you want. Should make the testing trivial

RyanRHall replied:
I'm going to push back again here. I don't think the juice is worth the squeeze.[ Here is an example](https://github.com/smartcontractkit/ccip/pull/1411/files) of what these changes would look like. It's a super marginal contract size/gas savings in the offramp but a more annoying test suite with a fair amount of code duplication b/t the tests and the implementation. IMHO this increases the overall code complexity.

I think maintaining some modularity for the sake of easier testing should be seen as an intentional feature in this case!

RayXpub commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -342,8 +337,6 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
     uint256[] memory manualExecGasLimits
   ) internal {
     uint64 sourceChainSelector = report.sourceChainSelector;
-    _whenNotCursed(sourceChainSelector);
```
q: why are we removing the curse check?

RensR replied:
!!!

RayXpub commented on contracts/src/v0.8/ccip/onRamp/OnRamp.sol:
```diff
@@ -466,8 +458,6 @@ contract OnRamp is IEVM2AnyOnRampClient, ITypeAndVersion, OwnerIsCreator {
     uint64 destChainSelector,
     Client.EVM2AnyMessage calldata message
   ) external view returns (uint256 feeTokenAmount) {
-    if (IRMN(i_rmnProxy).isCursed(bytes16(uint128(destChainSelector)))) revert CursedByRMN(destChainSelector);
```
q: same as https://github.com/smartcontractkit/ccip/pull/1360/files#r1735562477

elatoskinas commented on contracts/scripts/native_solc_compile_all_ccip:
```diff
@@ -59,6 +59,7 @@ compileContract () {
 # Contracts should be ordered in reverse-import-complexity-order to minimize overwrite risks.
 compileContract ccip/offRamp/EVM2EVMOffRamp.sol
 compileContract ccip/offRamp/OffRamp.sol
+compileContract ccip/rmn/RMNRemote.sol
```
q: do we need to include `RMNHome` as well?

RensR replied:
Yes, we need to use both in offchain tests

RyanRHall replied:
Yes, saving that for follow up PR :)

elatoskinas commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -148,8 +143,8 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
-  /// @dev The address of the RMN proxy
-  address internal immutable i_rmnProxy;
+  /// @dev The RMN verification contract
+  IRMNRemote internal immutable i_rmnRemote;
```
I think we should still use the RMN Proxy, because we will need one for the Router, which means we will have to point the proxy to the new RMN contract anyway.

This name might cause some confusion that it should point to the real `IRMNRemote` contract. Should we name this as `i_rmnRemoteProxy`?

RyanRHall replied:
Bit of a bikeshed here, but i feel like the OffRamp doesn't need to know or care if it is talking to the RMNRemote, a proxy, or a mock. It only cares that the dependency satisfies the `IRMNRemote` interface. What actually sits behind that interface is more of an implementation / deployment detail left up to the deployer. Ex we use the proxy in production and in integration tests, but a mock with no proxy in unit tests.

> This name might cause some confusion that it should point to the real IRMNRemote

More bikeshedding - but I feel like if we use a contract type like `RMNRemote` that says - _the thing stored here is that exact contract_. But because we're using the interface `IRMNRemote` we're saying _I don't know what actually lives here, I just know it has these X functions._

The USDC token also uses a proxy, but if we were directly integrating with that token we wouldn't write `s_USDCProxy`, yeah?

elatoskinas replied:
> but i feel like the OffRamp doesn't need to know or care if it is talking to the RMNRemote, a proxy, or a mock

Definitely agree with this, having `IRMNRemote` as the type makes sense to me

> The USDC token also uses a proxy, but if we were directly integrating with that token we wouldn't write s_USDCProxy, yeah?

I think yes, but this case is a bit different, because for all RMN instances, we always use a proxy, and never the real contract. Not naming it specifically as a proxy might cause some tooling / integration confusion (e.g. there might be an assumption that we must always set it to the real contract, skipping over the proxy). If we have a robust mechanism on the tooling side then it makes sense

RensR replied:
Hmm ok I can get onboard with the non-proxy name. I do wonder if RMNRemote is always correct? Don't we have an RMNHome on the home chain for this field?

RyanRHall replied:
@RensR good call on `RMNHome` vs `RemoteRemote`. Maybe just `RMN` then?

RyanRHall replied:
@elatoskinas I think you raise a good concern around deployments. I actually did the exact thing you were concerned about in an earlier draft of this PR - I confused the proxy and the implementation in a deployment script and committed the wrong thing without breaking any tests. Though I messed it up even with the word "proxy" staring me in the face lol.

I wonder if we might benefit from a dedicated test suite for deployments that could be shared between integration tests and _actual_ deployments? I feel like that might give us the strongest assurances against mis-configurations. Something that could do like `assert(offRamp.getRMN().typeAndVersion().matches("RMN Proxy"))` cc @connorwstein

elatoskinas replied:
Yeah a validation at the tooling side makes sense, we could do either a `typeAndVersion`, or we might have the reference to the RMN Proxy already, so we could just use that.

Not sure if we can re-use the deployment scripts directly though since it might not live on the same codebase / tool, maybe @asoliman92 or @AnieeG would have more context

elatoskinas commented on contracts/src/v0.8/ccip/rmn/RMNHome.sol:
```diff
@@ -3,7 +3,7 @@ pragma solidity ^0.8.24;
+import {ITypeAndVersion} from "../../shared/interfaces/ITypeAndVersion.sol";
 
 /// @notice Stores the home configuration for RMN, that is referenced by CCIP oracles, RMN nodes, and the RMNRemote
 /// contracts.
```
I had some open questions in the #1308 PR for this contract. My main question would be - shouldn't RMNHome be a capability in the Capability Registry, as we manage nodes and DONs in the Cap Registry? This way we do not need to duplicate any node management / version logic, and we are uniformly integrated to Keystone's model.

The remaining questions in that PR were on performance

RyanRHall replied:
We opted to merge that PR without addressing any feedback to unblock this one. After this is merged I'm going to work on productionizing `RMNRemote` & `RMNHome` and we can have this discussion then?

elatoskinas replied:
Sounds good

elatoskinas commented on contracts/src/v0.8/ccip/rmn/RMNRemote.sol:
```diff
@@ -1,14 +1,15 @@
 // SPDX-License-Identifier: BUSL-1.1
```
I had some open questions / callouts in #1308 for this contract

RyanRHall replied:
Same response as above

elatoskinas commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -583,6 +576,11 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
 
+    // Verify RMN signatures
+    if (commitReport.merkleRoots.length > 0) {
+      i_rmnRemote.verify(commitReport.merkleRoots, commitReport.rmnSignatures);
```
q: is it safe enough to only verify once in commit, and never re-verify blessed state via `i_rmnRemote` in exec? What happens if a root was wrongly committed? (full redeployment of RMN and ramps?)

From my understanding, we don't want the ability to remove blessed roots (which is consistent with the previous behaviour). How about lane cursing / global cursing in the case of a critical condition such as a wrongly committed root?

RensR replied:
afaik we just commit blessed roots and because they are already approved by two independent systems they are good. @gtklocker can you confirm?

RyanRHall replied:
> is it safe enough to only verify once in commit, and never re-verify blessed state via i_rmnRemote in exec?

In a way, we still are re-verifying that a commit is blessed in exec because we check that the commit _exists_. In this new model, a commit cannot exist on-chain _unless_ it is blessed. So by checking that a commit is present, you are also guaranteed that it is blessed. If you were to re-verify in exec, you would pass in the same data and sigs that were already verified during commit.

> What happens if a root was wrongly committed? (full redeployment of RMN and ramps?)

This can't happen without a break in the security model shared by commit DON and RMN. If a wrong root gets committed here, it is the same as a wrong root being committed and then blessed in v1.5, which we consider infeasible and unrecoverable. Ex in 1.5 we don't allow manually pruning commits that were already blessed; those are there for good.

RensR commented on contracts/src/v0.8/ccip/interfaces/IRMNRemote.sol:
```diff
@@ -0,0 +1,29 @@
+
+/// @dev Struct to hold a merkle root and an interval for a source chain so that an array of these can be passed in the CommitReport.
+/// @dev RMN depends on this struct, if changing, please notify the RMN maintainers.
+struct MerkleRoot {
```
Let's move the structs inside the interface

RyanRHall replied:
We could also move to a shared file? It doesn't really "belong" to either the `OffRamp` nor the `RMNRemote`, and I think it looks clunky to reference `RMNRemote.MerkleRoot` everywhere in the OffRamp. Not a strong opinion.

RensR replied:
Oh if it's shared between internal contracts it should go to Internal :)

RyanRHall replied:
I'm going to cave on this one :) But! I do want to say I think the `Internal` library has become a bit of a dumping ground and I'd love to see us migrate away from using it in this way. I think we can decrease both verbosity and compile times by moving to explicit imports.

RensR replied:
I agree that it having 1.5 and 1.6 structs makes it kinda not great. Let's see what happens once we move away from 1.5

RensR commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -8,7 +8,7 @@ import {IFeeQuoter} from "../interfaces/IFeeQuoter.sol";
 import {INonceManager} from "../interfaces/INonceManager.sol";
 import {IPoolV1} from "../interfaces/IPool.sol";
-import {IRMN} from "../interfaces/IRMN.sol";
+import {IRMNRemote, MerkleRoot, Signature as RMNSig} from "../interfaces/IRMNRemote.sol";
```
Not sure I love renaming. If we move the structs inside the interface that won't be needed right?

RyanRHall replied:
I put Sig inside the interface :)

RensR commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -83,9 +82,12 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
   struct StaticConfig {
     uint64 chainSelector; // ───╮  Destination chainSelector
-    address rmnProxy; // ───────╯  RMN proxy address
+    IRMNRemote rmnRemote; // ───╯  RMN Verification Contract
```
It's still the proxy right? Would be more clear if we kept the proxy naming

RyanRHall replied:
https://github.com/smartcontractkit/ccip/pull/1360#discussion_r1736635533

RensR commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -605,15 +603,12 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
       MerkleRoot memory root = commitReport.merkleRoots[i];
       uint64 sourceChainSelector = root.sourceChainSelector;
-
-      _whenNotCursed(sourceChainSelector);
```
Why do we remove this?

RyanRHall committed:
Revert "remove cursing from OffRamp"

This reverts commit c6469b316acc1080d3b36a45ddf6cc12592d034d.

RyanRHall committed:
Revert "remove RMN from OnRamp"

This reverts commit a4d71354a5cc3558cae654898ea39f82a42d9813.

RyanRHall committed:
update OnRamp to use RMNRemote

RyanRHall committed:
make PR suggested changes

RyanRHall committed:
Revert "upgrade solhint"

This reverts commit 1fc49367f88b0ae7a952a1c4d1e2e1abcc914190.

RensR commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -118,38 +121,25 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
 
   /// @notice a sequenceNumber interval
-  /// @dev RMN depends on this struct, if changing, please notify the RMN maintainers.
   struct Interval {
```
Now only used in events/errors, we can remove the struct

RyanRHall replied:
yeah good call 👍

RensR commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -83,9 +83,12 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
   struct StaticConfig {
     uint64 chainSelector; // ───╮  Destination chainSelector
-    address rmnProxy; // ───────╯  RMN proxy address
+    IRMNRemote rmnRemote; // ───╯  RMN Verification Contract
```
In CCIP we generally use basic types here, so address over IRMNRemote. It makes no difference for offchain and makes tests cleaner. Even if just for consistency I'd like this to be an address as well

RyanRHall replied:
I usually lean the opposite way on this one.

The big reason is I think it makes code cleaner & safer (no casting required). I think it's safer to define something's type once rather than define something's type every time we use it.

The other advantage I see is better guardrails for engineers. The type safety give you more immediate feedback of mistakes (usually via a red squiggly line in your editor) or if you miss that, a compiler error that fails fast/loud. If everything is just an address, you have to wait for a runtime error in a test and hope that you have a test that catches the issue!

It's also in our style guide 🙃.

RensR commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -83,9 +83,12 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
 
   /// @notice Struct that contains the static configuration
   /// @dev RMN depends on this struct, if changing, please notify the RMN maintainers.
+  /// @dev not sure why solhint complainins about this, seems like a buggy detector
```
typo complainins

RensR commented on contracts/src/v0.8/ccip/rmn/RMNRemote.sol:
```diff
@@ -1,14 +1,15 @@
+import {ITypeAndVersion} from "../../shared/interfaces/ITypeAndVersion.sol";
+import {IRMNRemote, MerkleRoot} from "../interfaces/IRMNRemote.sol";
 
 bytes32 constant RMN_V1_6_ANY2EVM_REPORT = keccak256("RMN_V1_6_ANY2EVM_REPORT");
```
let's move this into the contract

RyanRHall replied:
DO NOT REVIEW THIS CONTRACT! It is out of scope in the PR and will be addressed in a follow up :)

RensR commented on contracts/src/v0.8/ccip/rmn/RMNRemote.sol:
```diff
@@ -1,14 +1,15 @@
+contract RMNRemote is OwnerIsCreator, ITypeAndVersion, IRMNRemote {
   /// @dev temp placeholder to exclude this contract from coverage
   function test() public {}
 
```
can we rename i_chainSelector to `i_localChainSelector`? also in the constructor arg

RensR commented on contracts/src/v0.8/ccip/rmn/RMNRemote.sol:
```diff
@@ -1,14 +1,15 @@
 
+/// @dev XXX DO NOT USE THIS CONTRACT, NOT PRODUCTION READY XXX
 /// @notice This contract supports verification of RMN reports for any Any2EVM OffRamp.
-contract RMNRemote is Ownable2Step, ITypeAndVersion {
```
need visibility modifiers for s_config, s_configCount and s_signers.

Can we also order the structs and storage like all other CCIP contracts?

RensR commented on contracts/src/v0.8/ccip/rmn/RMNRemote.sol:
```diff
@@ -146,6 +135,17 @@ contract RMNRemote is Ownable2Step, ITypeAndVersion {
     if (numSigners < s_config.minSigners) revert ThresholdNotMet();
```
Do we really have to count numSigners? Isn't it always `signatures.length` unless it reverts in the loop? That means we can move the `ThresholdNotMet` check way up before expensive storage reads in the loop

RensR commented on contracts/src/v0.8/ccip/rmn/RMNRemote.sol:
```diff
@@ -146,6 +135,17 @@ contract RMNRemote is Ownable2Step, ITypeAndVersion {
+  }
+
   ///
   /// Events
```
Let's move these up to be in line with all the other contracts

RensR commented on contracts/src/v0.8/ccip/test/BaseTest.t.sol:
```diff
@@ -84,7 +87,25 @@ contract BaseTest is Test {
+    // setup mock RMN & RMNRemote
     s_mockRMN = new MockRMN();
+    s_mockRMNRemote = IRMNRemote(makeAddr("MOCK RMN REMOTE"));
+    vm.etch(address(s_mockRMNRemote), bytes("fake bytecode"));
```
Do we need this? Wouldn't mocking the expected calls to a real contract do the job?

RyanRHall replied:
Foundry docs explain why we have to do this in the first "Note" section [here](https://book.getfoundry.sh/cheatcodes/mock-call).

I tend to prefer using foundry's `mockCall` for small dependancies like this just because I think think it's less code to manage, but I could be convinced to convert it to a full-sized stubbed-out contract like the [MockRMN](https://github.com/smartcontractkit/ccip/blob/ccip-develop/contracts/src/v0.8/ccip/test/mocks/MockRMN.sol).

RensR replied:
Ahh ok I didn't know it needed bytecode, cool. Makes sense

RensR commented on contracts/src/v0.8/ccip/test/e2e/MultiRampsEnd2End.sol:
```diff
@@ -3,8 +3,10 @@ pragma solidity 0.8.24;
 import {LockReleaseTokenPool} from "../../pools/LockReleaseTokenPool.sol";
 import {TokenAdminRegistry} from "../../tokenAdminRegistry/TokenAdminRegistry.sol";
+
 import "../helpers/MerkleHelper.sol";
```
nit; rm newline

RyanRHall committed:
move MerkleRoot to Internal lib

RyanRHall committed:
update plugin with new report structure

RyanRHall committed:
Merge branch 'ccip-develop' of github.com:smartcontractkit/ccip into CCIP-2455-off-chain-blessing-off-ramp-contract-changes

RyanRHall committed:
forge fmt

makramkd committed:
remove bootstrapP2PIds (#1388)

## Motivation

Bootstrap P2P IDs are no longer needed in the OCR config in CCIPConfig.
See #1348's description for more details.

## Solution

Remove bootstrap P2P IDs from the OCR config in CCIPConfig.

## Related PRs
https://github.com/smartcontractkit/chainlink-ccip/pull/89

RyanRHall committed:
Merge branch 'ccip-develop' of github.com:smartcontractkit/ccip into CCIP-2455-off-chain-blessing-off-ramp-contract-changes

RyanRHall committed:
rename IRMNRemote -> IRMNV2

RyanRHall committed:
Merge branch 'ccip-develop' of github.com:smartcontractkit/ccip into CCIP-2455-off-chain-blessing-off-ramp-contract-changes

RyanRHall committed:
update wrappers & snapshot

makramkd commented on contracts/src/v0.8/ccip/offRamp/OffRamp.sol:
```diff
@@ -117,40 +120,20 @@ contract OffRamp is ITypeAndVersion, MultiOCR3Base {
-    uint64 sourceChainSelector; // Remote source chain selector that the Merkle Root is scoped to
-    bytes32 merkleRoot; // Merkle root of a single remote source chain
+    Internal.MerkleRoot[] merkleRoots; // Collection of merkle roots per source chain to commit
+    IRMNV2.Signature[] rmnSignatures; // RMN signatures on the merkle roots
```
If we want to be accurate these aren't signatures on the merkle roots exactly. They're signatures on a particular RMN report that consists of "lane updates" that match the updates in the `merkleRoots` array above. Maybe reference RMNRemote.verify() for folks who want to see the exact message being signed.

RyanRHall replied:
you're right. If I make additional changes, I'll fix the comment :)

makramkd commented on contracts/src/v0.8/ccip/rmn/RMNRemote.sol:
```diff
@@ -85,35 +87,23 @@ contract RMNRemote is Ownable2Step, ITypeAndVersion {
     return VersionedConfig({version: s_configCount, config: s_config});
   }
 
-  /// @notice The part of the LaneUpdate for a fixed destination chain and OffRamp, to avoid verbosity in Report
-  struct DestLaneUpdate {
-    uint64 sourceChainSelector;
-    bytes onrampAddress; // generic, to support arbitrary sources; for EVM2EVM, use abi.encodePacked
-    uint64 minMsgNr;
-    uint64 maxMsgNr;
-    bytes32 root;
```
I don't think we want to delete this, RMN will be signing a payload that includes this struct.

RyanRHall replied:
We discussed that with research & RMN teams [here](https://chainlink-core.slack.com/archives/C02PBKHCP28/p1724267938984369). We wanted to sync on the data structure so that commit DON & RMN can use the same struct.

makramkd replied:
Ah yeah you're right, they're identical just fields ordered differently to pack better.

makramkd commented on contracts/src/v0.8/ccip/interfaces/IRMNV2.sol:
```diff
@@ -0,0 +1,22 @@
+    bytes32 s;
+  }
+
+  function verify(Internal.MerkleRoot[] memory merkleRoots, Signature[] memory signatures) external view;
```
This function signature looks different than what is in the RMN design doc

RyanRHall replied:
same as above. We got sign off from RMN team that they're willing to change it. Please confirm @gtklocker

makramkd replied:
Ah dang nvm I misread things, my bad.

makramkd commented on contracts/src/v0.8/ccip/interfaces/IRMNV2.sol:
```diff
@@ -0,0 +1,22 @@
+    bytes32 s;
+  }
+
+  function verify(Internal.MerkleRoot[] memory merkleRoots, Signature[] memory signatures) external view;
```
A basic docstring on the expected behavior of this function might be useful (e.g maybe for implementing mocks), i.e "verifies RMN signatures against provided MerkleRoot structs and reverts on invalid signatures"
